## 1. Introduction

For a new self-driving car, we need an implementation of a high-precision pi: ChatGPT v4 suggests the following implementation for computing pi in Python, including a unit test. The code is packed in the two files `estimate_pi.py` and `test_estimate_pi.py`. 

Run the existing test using your shell (every cell starting with an `!` will be executed in your OS's shell). 

In [1]:
!python estimate_pi.py 
!python test_estimate_pi.py

3.1419


.
----------------------------------------------------------------------
Ran 1 test in 0.926s

OK


What is problematic with that test ChatGPT created for us and would you address this problem? (**1 point**)

    The problem is the function used in row 8 in test_estimate_pi.py; "self.assertAlmostEqual(pi_expected, pi_actual, delta=0.01).


    Problem.
    The test will always return OK as suggested by the documentation of the function. that is because of the asserAlmostEqual() in row 8 in the test. it will say that the test is "OK" if the difference is less than delta=0.01. meaning that the error have to be very large for the difference to be MORE then 0.01.
    
    Solution.
    1. use a lower delta. to get a more accurate estimation
    2. experiment with other Tests

## 2. Coverage

### 2.1 Statement coverage
Compute the statement coverage of the program using [`coverage.py`](https://coverage.readthedocs.io/en/latest/index.html). (**2 points**)

In [2]:
!coverage run test_estimate_pi.py
!coverage report -m

.
----------------------------------------------------------------------
Ran 1 test in 2.181s

OK


Name                  Stmts   Miss  Cover   Missing
---------------------------------------------------
estimate_pi.py           26     12    54%   18-19, 23-36
test_estimate_pi.py       9      0   100%
---------------------------------------------------
TOTAL                    35     12    66%


How can we interprete the results?

    Interpretation

    Stmts: defined as any executable line of code that performs an action, such as variable assignments, function calls, or control flow statements.

    Miss: What statements that the program has not executed.

    Cover: percentage of the code that have been executed by the program.

    Missing: the Rows that has not been executed.

    100% coverage in test_estimate_pi.py. meaning that all the Statements in "test_estimate_pi.py" runs.

    54% coverage in "estimate_pi.py" because our single test in "test_estimate_pi.py" only runs/calls a fraction of the code/statements in "estimate_pi.py".

    26 Total statments in estimate_pi.py, 12 of them were not exectued. the 12 rows that were missing are: 18-19, 23-36

    Lastly we have the "total" row where it summerize the statements, miss, coverage and missing across the files that we ran.

### 2.2 Branch coverage
Now compute the branch coverage of the program using [`coverage.py`](https://coverage.readthedocs.io/en/latest/index.html). (**2 points**)

In [3]:
!coverage run --branch test_estimate_pi.py
!coverage report -m

.
----------------------------------------------------------------------
Ran 1 test in 2.455s

OK


Name                  Stmts   Miss Branch BrPart  Cover   Missing
-----------------------------------------------------------------
estimate_pi.py           26     12     10      1    53%   18-19, 23-36
test_estimate_pi.py       9      0      2      1    91%   43->exit
-----------------------------------------------------------------
TOTAL                    35     12     12      2    62%


How can we interprete the results?

    Interpretation

    We could interprete branch coverage as how much of the logical paths that has been covered, not just how many lines, but also how it is covered in each statements

    Stmts: defined as any executable line of code that performs an action, such as variable assignments, function calls, or control flow statements.

    Miss: What statements that the program has not executed.

    Branch: example in an if-else statement, we have 2 branches, and if we only have gone in to the if branch and not the else, we have 2 branches but only 1 coverd. in Branch coverage we try to check all possible outcomes that the program can give.

    BrPart: How many parts of the branches we ahve found, are not checked.

    Cover: percentage of the code that have been executed by the program.

    Missing: the Rows that has not been executed.

    91% coverage in test_estimate_pi.py. meaning that some Statements in "test_estimate_pi.py" runs.

    53% coverage in "estimate_pi.py" because our single test in "test_estimate_pi.py" only runs/calls a fraction of the code/statements in "estimate_pi.py". Why it droped form 54% to 53%, is because we now ahve more things we check for. Mainly beacause the program jumps over 1 possible path "BrPart" in a branch and this is the cause of the decrease.

    26 Total statments in estimate_pi.py, 12 of them were not exectued. the 12 rows that were missing are: 18-19, 23-36. additionally we have 10 branches, and 1 branch that was only partially executed.

    test_estimate_pi.py have 2 branches. 1 branch was partially executed.

    Lastly we have the "total" row where it summerize the statements, miss, coverage and missing across the files that we ran.


## 3. More unit tests

Add two more unit tests with the principles you learned in the lecture. Describe what principle you have used. (**3 points**)


    Test1: test_zero_boundary
        principles: Self validating, Repeatable, independent.
        We check if the function returns a ZeroDivisionError when we call estimate_pi(). We have directly targeted the weakness in the function which is 0. Its independent because it does not depend on any other tests. Its repeatable beacuse its not dependent on external envioment. its self validating beacause it tells you when its passed or not.

    Test2: test_returns_float
        principles: Self validating, fast, repeatable, independent
        Checks if the type is a float. its self validating beacuse it tells you if its a float OR not. its fast beacuse we only pass 10 itterations to the function. its repeatable beacuse it does not depend on external enviroment. its independent beacuse we dont need to run anything before this test in order for the test to execute.

## 4. Mocking

We want to store the resulting number persistently on our file system. We use the following class. 

In [4]:
class PiFileWriter:
    @staticmethod
    def write(content, file_path):
        with open(file_path, 'w') as file:
            file.write(content)

Implement a test double for `PiFileWriter` and add your implementation to `test_estimate_pi.py`. Discuss what type of test double you have implemented. (**3 points**)

    we have implemented a test double of type "FAKE". we create a class that represent a .txt file with Variables written_content and written_path. we created a static method that writes and stores the informaton in the variables. Because a test double are supposed to be independent, we have chosen to hard code the pi_value. 

Name three other types of unit tests you would want to mock and explain why. (**1 point**)

    # 1
    * Logger and Sensor. Sensor takes time to return a value, so we MOCK the sensor that has pre recorded values that are similar to the sensor. This will give us a return instantly for our test.

    # 2
    * Register users on a webpage. We want to test if the username registration works correctly, instead of running the webserver all the time, or request information all the time, we can MOCK the usernames, so that the MOCK sends us randomized or hardcorded usernames. This will save resources and time. Also we can MOCK confirmation emails.

    # 3
    * Borrowing school books. You MOCK the system that keeps track of how long you are able to loan a book. The MOCK in this case will return a specified (hardcoded) date. 

## 5. Coverage revisited

Rerun statement and branch coverage and discuss the differences and changes. (**3 points**)

In [4]:
!coverage run test_estimate_pi.py
!coverage report -m


!coverage run --branch test_estimate_pi.py
!coverage report -m

....
----------------------------------------------------------------------
Ran 4 tests in 2.199s

OK


Name                  Stmts   Miss  Cover   Missing
---------------------------------------------------
estimate_pi.py           26     12    54%   18-19, 23-36
test_estimate_pi.py      28      0   100%
---------------------------------------------------
TOTAL                    54     12    78%


....
----------------------------------------------------------------------
Ran 4 tests in 2.402s

OK


Name                  Stmts   Miss Branch BrPart  Cover   Missing
-----------------------------------------------------------------
estimate_pi.py           26     12     10      1    53%   18-19, 23-36
test_estimate_pi.py      28      0      2      1    97%   43->exit
-----------------------------------------------------------------
TOTAL                    54     12     12      2    73%



TA BORT SEN GLÖM IIIIIIIIIIIIIIIIINNNNNNNNNNNNNNNNNNNNNTTTTTTTTTTTTTTTTTTTTTTTTEEEEEEEEEEEEEEEEEEEEEEEE


statment Name                  Stmts   Miss  Cover   Missing
---------------------------------------------------
estimate_pi.py           26     12    54%   18-19, 23-36
test_estimate_pi.py       9      0   100%
---------------------------------------------------
TOTAL                    35     12    66%


branchName                  Stmts   Miss Branch BrPart  Cover   Missing
-----------------------------------------------------------------
estimate_pi.py           26     12     10      1    53%   18-19, 23-36
test_estimate_pi.py       9      0      2      1    91%   46->exit
-----------------------------------------------------------------
TOTAL                    35     12     12      2    62%


--------------------------------------------------------------

OK
Name                  Stmts   Miss  Cover   Missing
---------------------------------------------------
estimate_pi.py           26     12    54%   18-19, 23-36
test_estimate_pi.py      28      0   100%
---------------------------------------------------
TOTAL                    54     12    78%
....
----------------------------------------------------------------------
Ran 4 tests in 2.402s

OK
Name                  Stmts   Miss Branch BrPart  Cover   Missing
-----------------------------------------------------------------
estimate_pi.py           26     12     10      1    53%   18-19, 23-36
test_estimate_pi.py      28      0      2      1    97%   43->exit
-----------------------------------------------------------------
TOTAL                    54     12     12      2    73%





    Why we wont get 100% covarage in estimate_pi is becasue we created a test double in test_estimate_file. so it will never run the Original class.

    Statement Coverage rose from 66% to 78% and Branch Coverage rose from 62% to 73%. We added more tests in the test file which leads to more coverage in total. the class PiFileWriter was never tested when we ran the coverage code in question 2. 

    the coverage in the file estimate_pi are unchanged to 54% statement and 53% branch coverage because we only tested things that were already in the file. 
    We created a test dubble which inherit the main function of the class PiFileWriter and thats why we still have the same coverage for those files.

    One difference is between branch coverages in test_estimate_pi-py beacause we have more tests in the final test file, we get more statements, but the branch and BrPart are still the same. this leads to us getting 97% branch coverage instead of 91% coverage as we had before we added our tests.

## 6. Submit to Canvas

Almost done, but the most tricky part is missing: submitting. :)

Before submitting, make sure
- you don't use external libraries except `coverage.py`
- the notebook runs straight through
- your test code works
- your code is readable and follows the Python coding conventions

- Submit these files to canvas:
    - `assignment_1.ipynb`
    - `test_estimate_pi.py`
